File: hybrid_roadmap/modules/roadmap/personalized_roadmap.py
"""
Personalized Job Roadmap Generator
Generates detailed, time-aware learning roadmaps for specific jobs.
Uses Gemini for intelligent subtopic generation.
Implements smart parallel learning for multiple missing skills.
"""

import os
import json
import re
from datetime import datetime, timedelta
from .llama_agent import generate_subtopics


def calculate_days_until_deadline(deadline_str):
    """
    Calculate days remaining until deadline.
    Expects ISO format date string or relative days.
    """
    try:
        if isinstance(deadline_str, (int, float)):
            return int(deadline_str)
        
        deadline = datetime.fromisoformat(deadline_str)
        days_left = (deadline - datetime.now()).days
        return max(1, days_left)
    except Exception as e:
        print(f"⚠️ Warning: Could not parse deadline '{deadline_str}', using default 30 days: {e}")
        return 30


def rank_skills_by_importance(matched_skills, missing_skills, jd_skill_weights=None):
    """
    Rank missing skills by importance based on:
    1. Frequency in JD
    2. Position in requirements (must-have vs nice-to-have)
    3. Complexity estimation
    """
    ranked = []
    
    for skill in missing_skills:
        skill_lower = skill.lower()
        
        # Get weight from JD if available
        weight = 0
        if jd_skill_weights:
            weight = jd_skill_weights.get(skill, jd_skill_weights.get(skill_lower, 0))
        
        # Default weights if not in JD
        if weight == 0:
            if skill_lower in ['python', 'java', 'javascript', 'sql', 'react', 'node', 'express']:
                weight = 3  # High priority core skills
            elif skill_lower in ['aws', 'docker', 'kubernetes', 'git', 'linux']:
                weight = 2  # Medium priority infrastructure
            else:
                weight = 1  # Lower priority specialized skills
        
        ranked.append({
            'skill': skill,
            'weight': weight,
            'complexity': estimate_complexity(skill)
        })
    
    # Sort by weight (descending) then by complexity
    ranked.sort(key=lambda x: (-x['weight'], -x['complexity']))
    return ranked


def estimate_complexity(skill):
    """
    Estimate learning complexity (1-5 scale).
    Returns hours needed for basic proficiency.
    """
    skill_lower = skill.lower()
    
    complexity_map = {
        # Very hard (5)
        'machine learning': 5, 'deep learning': 5, 'tensorflow': 5, 'pytorch': 5,
        'kubernetes': 5, 'distributed systems': 5, 'system design': 5,
        
        # Hard (4)
        'aws': 4, 'azure': 4, 'gcp': 4, 'devops': 4, 'microservices': 4,
        'docker': 4, 'data engineering': 4, 'apache spark': 4,
        
        # Medium (3)
        'react': 3, 'node.js': 3, 'express': 3, 'sql': 3, 'mongodb': 3,
        'python': 3, 'java': 3, 'javascript': 3, 'rest api': 3,
        
        # Easy (2)
        'html': 2, 'css': 2, 'git': 2, 'linux': 2, 'npm': 2, 'yarn': 2,
        
        # Very easy (1)
        'communication': 1, 'problem solving': 1, 'teamwork': 1,
    }
    
    for key, value in complexity_map.items():
        if key in skill_lower:
            return value
    
    return 3  # Default medium complexity


def distribute_skills_by_time(ranked_skills, days_available):
    """
    Intelligently distribute skills across available time.
    Returns list of skill groups with timing info.
    
    Strategy:
    - If 1-2 skills: sequential focus
    - If 3-4 skills: 2 parallel + 1 sequential
    - If 5+ skills: 2-3 parallel streams
    """
    if not ranked_skills or days_available < 1:
        return []
    
    num_skills = len(ranked_skills)
    
    # Determine parallel streams
    if num_skills == 0:
        return []
    elif num_skills == 1:
        parallel_count = 1
    elif num_skills == 2:
        parallel_count = 2
    elif num_skills <= 4:
        parallel_count = 2  # 2 parallel, rest sequential
    else:
        parallel_count = 3  # 3 parallel streams for many skills
    
    # Distribute skills across parallel streams (round-robin)
    streams = [[] for _ in range(parallel_count)]
    for i, skill in enumerate(ranked_skills):
        streams[i % parallel_count].append(skill)
    
    # Calculate timing for each skill
    result = []
    skill_index = 0
    
    for stream_idx, stream in enumerate(streams):
        # Days per stream = total days / number of streams
        stream_days = max(1, days_available // parallel_count)
        
        for skill_idx, skill in enumerate(stream):
            # Within a stream, distribute time proportionally to complexity
            estimated_hours = skill['complexity'] * 10  # rough estimate: complexity * 10 hours
            estimated_days = max(2, estimated_hours // 4)  # Assuming 4 hours/day learning
            
            result.append({
                'skill': skill['skill'],
                'weight': skill['weight'],
                'complexity': skill['complexity'],
                'parallel_stream': stream_idx + 1,
                'stream_position': skill_idx + 1,
                'total_stream_days': stream_days,
                'estimated_days': min(estimated_days, stream_days),
                'order': skill_index
            })
            skill_index += 1
    
    return result


def generate_skill_roadmap(skill, days_available, parallel_info=None):
    """
    Generate detailed learning roadmap for a single skill.
    Returns subtopics with estimated duration and resources.
    """
    try:
        subtopics = generate_subtopics(skill)
    except Exception as e:
        print(f"⚠️ AI generation failed for {skill}, using fallback: {e}")
        subtopics = [
            f"{skill} Fundamentals & Basics",
            f"{skill} Core Concepts & Architecture",
            f"{skill} Practical Implementation",
            f"{skill} Advanced Patterns & Optimization",
            f"{skill} Real-world Projects"
        ]
    
    # Distribute days among subtopics
    estimated_days = parallel_info.get('estimated_days', days_available // 5) if parallel_info else days_available // 5
    days_per_subtopic = max(1, estimated_days // len(subtopics))
    
    roadmap = {
        "skill": skill,
        "estimated_days": estimated_days,
        "parallel_stream": parallel_info.get('parallel_stream', 1) if parallel_info else 1,
        "complexity": parallel_info.get('complexity', 3) if parallel_info else 3,
        "subtopics": []
    }
    
    for i, topic in enumerate(subtopics):
        subtopic = {
            "title": topic,
            "estimated_days": days_per_subtopic,
            "week": (i // 2) + 1,  # Group into weeks (2 subtopics per week typically)
            "resources": {
                "youtube": [
                    f"https://www.youtube.com/results?search_query={skill.replace(' ', '+')}+{topic.replace(' ', '+')}+tutorial"
                ],
                "docs": [
                    f"https://www.google.com/search?q={skill.replace(' ', '+')}+{topic.replace(' ', '+')}+documentation"
                ],
                "github": [
                    f"https://github.com/search?q={skill.replace(' ', '+')}+{topic.replace(' ', '+')}+project"
                ],
                "practice": [
                    f"https://www.youtube.com/results?search_query={skill.replace(' ', '+')}+{topic.replace(' ', '+')}+practice"
                ]
            },
            "practice_projects": [
                f"Build a mini-project demonstrating {topic}",
                f"Solve 5-10 practice problems on {topic}",
                f"Create a portfolio piece showcasing {topic}"
            ],
            "milestones": [
                f"Understand core concepts of {topic}",
                f"Complete practice exercises",
                f"Build project using {topic}"
            ]
        }
        roadmap["subtopics"].append(subtopic)
    
    return roadmap


def build_personalized_job_roadmap(resume_skills, jd_skills, jd_skill_weights=None, deadline_str=None, days_fallback=30):
    """
    Master function: Build comprehensive personalized job roadmap.
    
    Args:
        resume_skills: List of skills student has
        jd_skills: List of skills required by job
        jd_skill_weights: Dict of skill -> importance weight from JD parsing
        deadline_str: ISO date string or days remaining until deadline
        days_fallback: Default days if deadline can't be parsed
    
    Returns:
        Comprehensive roadmap with all details, timing, and resources
    """
    # Calculate available time
    days_available = calculate_days_until_deadline(deadline_str) if deadline_str else days_fallback
    
    # Identify matched and missing skills
    resume_set = set(s.lower() for s in resume_skills)
    jd_set = set(s.lower() for s in jd_skills)
    
    matched = list(resume_set.intersection(jd_set))
    missing = [skill for skill in jd_skills if skill.lower() not in resume_set]
    
    # Rank missing skills by importance
    ranked_missing = rank_skills_by_importance(matched, missing, jd_skill_weights)
    
    # Distribute skills intelligently across available time
    skill_distribution = distribute_skills_by_time(ranked_missing, days_available)
    
    # Generate detailed roadmap for each missing skill
    skill_roadmaps = []
    for skill_info in skill_distribution:
        skill_roadmap = generate_skill_roadmap(
            skill_info['skill'],
            skill_info['estimated_days'],
            skill_info
        )
        skill_roadmaps.append(skill_roadmap)
    
    # Build comprehensive response
    response = {
        "status": "success",
        "analysis": {
            "resume_skills_count": len(resume_skills),
            "jd_skills_count": len(jd_skills),
            "matched_skills": matched,
            "matched_count": len(matched),
            "missing_skills": missing,
            "missing_count": len(missing),
            "match_percentage": round((len(matched) / len(jd_skills) * 100), 2) if jd_skills else 0,
            "days_available": days_available
        },
        "timeline": {
            "start_date": datetime.now().isoformat(),
            "target_completion": (datetime.now() + timedelta(days=days_available)).isoformat(),
            "total_days": days_available,
            "learning_hours_per_day": 4,  # Recommended
            "total_learning_hours": sum(s['estimated_days'] * 4 for s in skill_roadmaps)
        },
        "learning_strategy": {
            "approach": f"Parallel learning with {len(set(s['parallel_stream'] for s in skill_roadmaps))} focus streams",
            "parallel_streams": len(set(s['parallel_stream'] for s in skill_roadmaps)),
            "sequential_order": False,
            "rationale": "Learn high-priority skills in parallel to maximize prep time while maintaining focus"
        },
        "skill_roadmaps": skill_roadmaps,
        "weekly_schedule_template": generate_weekly_template(skill_roadmaps, days_available),
        "success_metrics": {
            "minimum_competency": "Basic understanding + 1 small project per skill",
            "target_competency": "Intermediate + portfolio-ready project per skill",
            "success_criteria": [
                "Complete 80% of subtopics for all skills",
                "Build 1 project per skill demonstrating competency",
                "Pass mock interviews covering missing skills",
                "Achieve 70%+ match with job requirements"
            ]
        }
    }
    
    return response


def generate_weekly_template(skill_roadmaps, total_days):
    """
    Generate a weekly learning schedule template.
    """
    weeks = max(1, total_days // 7)
    days_per_week = total_days / weeks if weeks > 0 else 7
    
    template = {
        "weeks": weeks,
        "days_per_week": round(days_per_week, 1),
        "sample_week_structure": {
            "monday": "Learn theory for Stream 1 skill",
            "tuesday": "Practice exercises for Stream 1 skill",
            "wednesday": "Learn theory for Stream 2 skill",
            "thursday": "Practice exercises for Stream 2 skill",
            "friday": "Build mini-project for any skill",
            "saturday": "Review + mock interview prep",
            "sunday": "Rest or catch up on pending topics"
        },
        "daily_routine": {
            "morning_2hours": "Theory & concepts learning",
            "afternoon_1.5hours": "Practice exercises & coding",
            "evening_0.5hours": "Documentation review & planning",
            "notes": "Adjust timing based on personal preferences"
        }
    }
    
    return template

File: server/utils/personalizedRoadmap.js
// personalizedRoadmap.js
require('dotenv').config();
const axios = require('axios');

// Using gemini-1.5-flash as it is the current standard stable flash model
const GEMINI_API_URL = 'https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent';

// Helper to calculate days until deadline
function calculateDaysUntilDeadline(deadlineStr) {
    try {
        if (!deadlineStr) return 30; // Default
        if (typeof deadlineStr === 'number') return deadlineStr;

        const deadline = new Date(deadlineStr);
        const now = new Date();
        const diffTime = deadline - now;
        const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));
        return Math.max(1, diffDays);
    } catch (e) {
        console.warn('Could not parse deadline, using default 30 days:', e);
        return 30;
    }
}

// Rank skills by importance
function rankSkillsByImportance(matchedSkills, missingSkills, jdSkillWeights = {}) {
    return missingSkills.map(skill => {
        let weight = jdSkillWeights[skill] || jdSkillWeights[skill.toLowerCase()] || 0;

        if (weight === 0) {
            const lower = skill.toLowerCase();
            if (['python', 'java', 'javascript', 'sql', 'react', 'node', 'express'].some(k => lower.includes(k))) weight = 3;
            else if (['aws', 'docker', 'kubernetes', 'git', 'linux'].some(k => lower.includes(k))) weight = 2;
            else weight = 1;
        }

        return {
            skill,
            weight,
            complexity: estimateComplexity(skill)
        };
    }).sort((a, b) => (b.weight - a.weight) || (b.complexity - a.complexity));
}

// Estimate complexity (simple heuristic)
function estimateComplexity(skill) {
    const lower = skill.toLowerCase();
    if (['machine learning', 'system design', 'kubernetes'].some(k => lower.includes(k))) return 5;
    if (['aws', 'microservices', 'docker'].some(k => lower.includes(k))) return 4;
    if (['react', 'node', 'sql', 'python', 'java'].some(k => lower.includes(k))) return 3;
    if (['html', 'css', 'git'].some(k => lower.includes(k))) return 2;
    return 3;
}

// Distribute skills logic
function distributeSkillsByTime(rankedSkills, daysAvailable) {
    if (!rankedSkills.length) return [];

    const numSkills = rankedSkills.length;
    let parallelCount = 1;
    if (numSkills === 2) parallelCount = 2;
    else if (numSkills <= 4) parallelCount = 2;
    else if (numSkills >= 5) parallelCount = 3;

    const result = [];
    const streams = Array.from({ length: parallelCount }, () => []);

    rankedSkills.forEach((item, i) => {
        streams[i % parallelCount].push(item);
    });

    streams.forEach((stream, streamIdx) => {
        const streamDays = Math.max(1, Math.floor(daysAvailable / parallelCount));
        stream.forEach((item, idx) => {
            const estimatedHours = item.complexity * 10;
            const estimatedDays = Math.ceil(estimatedHours / 4); // 4 hours/day

            result.push({
                ...item,
                parallel_stream: streamIdx + 1,
                total_stream_days: streamDays,
                estimated_days: Math.min(estimatedDays, streamDays)
            });
        });
    });

    return result;
}

// Generate subtopics using Gemini
async function generateSubtopicsWithGemini(skill) {
    const apiKey = process.env.GEMINI_API_KEY;
    if (!apiKey) {
        console.error('CRITICAL: GEMINI_API_KEY is missing from process.env');
        throw new Error('GEMINI_API_KEY not set');
    }

    const prompt = `Generate a structured learning path for the skill "${skill}".
  Return a JSON object with this EXACT structure:
  {
    "subtopics": [
      {
        "title": "Subtopic Title",
        "project": "Mini project idea description",
        "youtube_search_query": "Search query for tutorials"
      }
    ],
    "final_projects": {
      "suggested": ["Project Idea 1", "Project Idea 2"],
      "github_search_queries": ["repo query 1", "repo query 2"]
    }
  }
  Generate 4-6 subtopics.`;

    try {
        const response = await axios.post(
            `${GEMINI_API_URL}?key=${apiKey}`,
            { contents: [{ parts: [{ text: prompt }] }] },
            { headers: { 'Content-Type': 'application/json' }, timeout: 15000 }
        );

        const text = response.data.candidates[0].content.parts[0].text;
        const jsonMatch = text.match(/\{[\s\S]*\}/);
        const jsonStr = jsonMatch ? jsonMatch[0] : text;
        return JSON.parse(jsonStr);
    } catch (e) {
        if (e.response) {
            console.error(`Gemini generation failed for ${skill}: Status ${e.response.status}`, JSON.stringify(e.response.data));
        } else {
            console.error(`Gemini generation failed for ${skill}:`, e.message);
        }
        // Fallback
        return {
            subtopics: [
                { title: `${skill} Fundamentals`, project: `Basic ${skill} app`, youtube_search_query: `${skill} tutorial` },
                { title: `${skill} Advanced Concepts`, project: `Advanced ${skill} system`, youtube_search_query: `${skill} advanced tutorial` }
            ],
            final_projects: {
                suggested: [`Complete ${skill} Application`],
                github_search_queries: [`${skill} project`]
            }
        };
    }
}

// Main function to build roadmap
async function buildPersonalizedJobRoadmap({ resumeSkills, jdSkills, jdSkillWeights, deadline, daysFallback }) {
    const daysAvailable = calculateDaysUntilDeadline(deadline) || daysFallback;

    const resumeSet = new Set(resumeSkills.map(s => s.toLowerCase()));
    const missing = jdSkills.filter(s => !resumeSet.has(s.toLowerCase()));
    const matched = jdSkills.filter(s => resumeSet.has(s.toLowerCase()));

    const rankedMissing = rankSkillsByImportance(matched, missing, jdSkillWeights);
    const distribution = distributeSkillsByTime(rankedMissing, daysAvailable);

    // Create roadmap object compatible with RoadmapViewer
    const roadmap = {};

    // Generate content for each skill in parallel
    await Promise.all(distribution.map(async (item) => {
        const content = await generateSubtopicsWithGemini(item.skill);

        // Transform to frontend format
        roadmap[item.skill] = {
            main_course: item.skill,
            duration_weeks: Math.ceil(item.estimated_days / 7),
            subtopics: content.subtopics.map(sub => ({
                title: sub.title,
                project: sub.project,
                youtube_links: [
                    `https://www.youtube.com/results?search_query=${encodeURIComponent(sub.youtube_search_query || sub.title + ' tutorial')}`
                ]
            })),
            final_projects: {
                suggested: content.final_projects?.suggested || [],
                github_references: (content.final_projects?.github_search_queries || []).map(q =>
                    `https://github.com/search?q=${encodeURIComponent(q)}`
                )
            }
        };
    }));

    if (Object.keys(roadmap).length === 0 && matched.length > 0) {
        // If no missing skills, maybe generate for advanced topics of matched skills?
        // For now, just return empty or success message?
        // The frontend expects keys.
    }

    return {
        status: 'success',
        roadmap
    };
}

module.exports = {
    buildPersonalizedJobRoadmap
};
